## Import Libraries

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

#1 Membaca data

In [ ]:
# import data
df = pd.read_excel('titanic.xlsx')
df.head()

survived                                             name     sex      age
0         1                    Allen, Miss. Elisabeth Walton  female  29.0000
1         1                   Allison, Master. Hudson Trevor    male   0.9167
2         0                     Allison, Miss. Helen Loraine  female   2.0000
3         0             Allison, Mr. Hudson Joshua Creighton    male  30.0000
4         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000

In [ ]:
df.tail()

survived                                     name     sex   age
495         1  Mallet, Mrs. Albert (Antoinette Magnin)  female  24.0
496         0        Mangiavacchi, Mr. Serafino Emilio    male   NaN
497         0               Matthews, Mr. William John    male  30.0
498         0                Maybery, Mr. Frank Hubert    male  40.0
499         0                McCrae, Mr. Arthur Gordon    male  32.0

In [ ]:
df.sample(5)

survived                                               name     sex   age
378         1                   Collyer, Miss. Marjorie "Lottie"  female   8.0
281         1  Stengel, Mrs. Charles Emil Henry (Annie May Mo...  female  43.0
131         1            Gibson, Mrs. Leonard (Pauline C Boeson)  female  45.0
114         0                     Fortune, Mr. Charles Alexander    male  19.0
107         1               Flegenheim, Mrs. Alfred (Antoinette)  female   NaN

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  500 non-null    int64  
 1   name      500 non-null    object 
 2   sex       500 non-null    object 
 3   age       451 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 15.8+ KB


Observation:

1. Data berisi 4 kolom dan 500 baris
2. Terdapat data kosong pada kolom "age"
3. Jenis tipe data sesuai dengan namanya


#2 Statistical Summary

In [ ]:
df.columns

Index(['survived', 'name', 'sex', 'age'], dtype='object')

In [ ]:
categoricals = ['name', 'sex']
numericals = ['survived', 'age']

In [ ]:
df[categoricals].describe()

name   sex
count                              500   500
unique                             499     2
top     Eustis, Miss. Elizabeth Mussey  male
freq                                 2   288

Observation:

1. "name" memiliki data yang sama
2. "sex" memiliki pilihan perempuan dan laki-laki, dengan 288 laki-laki sisanya perempuan

In [ ]:
df[numericals].describe()

survived         age
count  500.000000  451.000000
mean     0.540000   35.917775
std      0.498897   14.766454
min      0.000000    0.666700
25%      0.000000   24.000000
50%      1.000000   35.000000
75%      1.000000   47.000000
max      1.000000   80.000000

Observation:

1. "age" menunjukkan distribusi simetris berdasarkan nilai rata-rata dan median
2. "survived" adalah kolom biner karena nilainya 0 atau 1

In [ ]:
for col in categoricals:
  print(f"Value counts of {col} column")
  print(df[col].value_counts(), '\n')

Value counts of name column
name
Eustis, Miss. Elizabeth Mussey                                                        2
Becker, Miss. Ruth Elizabeth                                                          1
Becker, Miss. Marion Louise                                                           1
Becker, Master. Richard F                                                             1
Beauchamp, Mr. Henry James                                                            1
Beane, Mrs. Edward (Ethel Clarke)                                                     1
Beane, Mr. Edward                                                                     1
Bateman, Rev. Robert James                                                            1
Banfield, Mr. Frederick James                                                         1
Ball, Mrs. (Ada E Hall)                                                               1
Baimbrigge, Mr. Charles Robert                                                        1

In [ ]:
for col in numericals:
  print(f"==== {col} ====")
  print(df[col].value_counts(), '\n')

==== survived ====
survived
1    270
0    230
Name: count, dtype: int64 

==== age ====
age
24.0000    23
30.0000    20
36.0000    19
18.0000    14
42.0000    14
45.0000    14
35.0000    14
22.0000    12
28.0000    12
23.0000    11
21.0000    11
31.0000    11
48.0000    11
25.0000    11
29.0000    10
54.0000    10
50.0000    10
34.0000    10
47.0000    10
39.0000     9
27.0000     9
33.0000     8
49.0000     8
38.0000     8
19.0000     8
40.0000     8
55.0000     8
32.0000     7
60.0000     7
17.0000     7
26.0000     7
44.0000     6
52.0000     6
37.0000     6
58.0000     6
64.0000     5
46.0000     5
57.0000     5
51.0000     5
16.0000     4
53.0000     4
56.0000     4
43.0000     4
62.0000     4
61.0000     4
1.0000      3
8.0000      3
63.0000     3
41.0000     3
15.0000     2
71.0000     2
20.0000     2
65.0000     2
4.0000      2
6.0000      2
2.0000      1
0.9167      1
14.0000     1
11.0000     1
76.0000     1
59.0000     1
80.0000     1
28.5000     1
32.5000     1
70.0000     

#3 Cek data duplikat


In [ ]:
len(df)

500

In [ ]:
len(df.drop_duplicates())

499

In [ ]:
len(df.drop_duplicates()) / len(df)

0.998

In [ ]:
list(df.columns)

['survived', 'name', 'sex', 'age']

In [ ]:
duplicates = df[df.duplicated(keep=False)]
duplicates

survived                            name     sex   age
104         1  Eustis, Miss. Elizabeth Mussey  female  54.0
349         1  Eustis, Miss. Elizabeth Mussey  female  54.0

In [ ]:
duplicates.groupby(list(df.columns)).size()

survived  name                            sex     age 
1         Eustis, Miss. Elizabeth Mussey  female  54.0    2
dtype: int64

In [ ]:
duplicates.groupby(list(df.columns)).size().reset_index(name='jumlah duplikat')

survived                            name     sex   age  jumlah duplikat
0         1  Eustis, Miss. Elizabeth Mussey  female  54.0                2

In [ ]:
# Langkah 2: Hitung frekuensi kemunculan tiap baris duplikat
duplicate_counts = duplicates.groupby(list(df.columns)).size().reset_index(name='jumlah_duplikat')

# Langkah 3: Urutkan berdasarkan jumlah duplikat
sorted_duplicates = duplicate_counts.sort_values(by='jumlah_duplikat', ascending=False)

# Tampilkan hasil
print("Baris duplikat yang sudah diurutkan berdasarkan jumlah kemunculannya:")
sorted_duplicates

Baris duplikat yang sudah diurutkan berdasarkan jumlah kemunculannya:


survived                            name     sex   age  jumlah_duplikat
0         1  Eustis, Miss. Elizabeth Mussey  female  54.0                2

In [ ]:
df = df.drop_duplicates()

In [ ]:
len(df.drop_duplicates()) / len(df)

1.0

#4 Cek missing value

In [ ]:
df.isna().sum()

survived     0
name         0
sex          0
age         49
dtype: int64

In [ ]:
for column in df.columns:
    print(f"============= {column} =================")
    display(df[column].value_counts())
    print()

============= survived =================


survived
1    269
0    230
Name: count, dtype: int64


============= name =================


name
McCrae, Mr. Arthur Gordon                                                             1
Allen, Miss. Elisabeth Walton                                                         1
Allison, Master. Hudson Trevor                                                        1
Lehmann, Miss. Bertha                                                                 1
Laroche, Mrs. Joseph (Juliette Marie Louise Lafargue)                                 1
Laroche, Mr. Joseph Philippe Lemercier                                                1
Laroche, Miss. Simonne Marie Anne Andree                                              1
Laroche, Miss. Louise                                                                 1
Lamb, Mr. John Joseph                                                                 1
Lahtinen, Rev. William                                                                1
Lahtinen, Mrs. William (Anna Sylfven)                                                 1
Kvillner, Mr. Johan Henrik Johannesson                                                1
Knight, Mr. Robert J                                                                  1
Kirkland, Rev. Charles Leonard                                                        1
Kelly, Mrs. Florence "Fannie"                                                         1
Keane, Mr. Daniel                                                                     1
Keane, Miss. Nora A                                                                   1
Karnes, Mrs. J Frank (Claire Bennett)                                                 1
Kantor, Mrs. Sinai (Miriam Sternin)                                                   1
Borebank, Mr. John James                                                              1
Bonnell, Miss. Elizabeth                                                              1
Bonnell, Miss. Caroline                                                               1
Blank, Mr. Henry                                                                      1
Blackwell, Mr. Stephen Weart                                                          1
Bjornstrom-Steffansson, Mr. Mauritz Hakan                                             1
Bissette, Miss. Amelia                                                                1
Bishop, Mrs. Dickinson H (Helen Walton)                                               1
Bishop, Mr. Dickinson H                                                               1
Birnbaum, Mr. Jakob                                                                   1
Bird, Miss. Ellen                                                                     1
Bidois, Miss. Rosalie                                                                 1
Behr, Mr. Karl Howell                                                                 1
Beckwith, Mrs. Richard Leonard (Sallie Monypeny)                                      1
Beckwith, Mr. Richard Leonard                                                         1
Beattie, Mr. Thomson                                                                  1
Cardeza, Mrs. James Warburton Martinez (Charlotte Wardle Drake)                       1
Cardeza, Mr. Thomas Drake Martinez                                                    1
Candee, Mrs. Edward (Helen Churchill Hungerford)                                      1
Calderhead, Mr. Edward Pennington                                                     1
Cairns, Mr. Alexander                                                                 1
Butt, Major. Archibald Willingham                                                     1
Burns, Miss. Elizabeth Margaret                                                       1
Bucknell, Mrs. William Robert (Emma Eliza Ward)                                       1
Brown, Mrs. John Murray (Caroline Lane Lamson)                                        1
Brown, Mrs. James Joseph (Margaret Tobin)                                             1
Brewe, Dr. Arthur Jackson          


============= sex =================


sex
male      288
female    211
Name: count, dtype: int64


============= age =================


age
24.0000    23
30.0000    20
36.0000    19
18.0000    14
42.0000    14
45.0000    14
35.0000    14
22.0000    12
28.0000    12
23.0000    11
21.0000    11
31.0000    11
48.0000    11
25.0000    11
29.0000    10
34.0000    10
50.0000    10
47.0000    10
39.0000     9
27.0000     9
54.0000     9
33.0000     8
49.0000     8
38.0000     8
19.0000     8
40.0000     8
55.0000     8
32.0000     7
60.0000     7
17.0000     7
26.0000     7
44.0000     6
52.0000     6
37.0000     6
58.0000     6
64.0000     5
46.0000     5
57.0000     5
51.0000     5
16.0000     4
53.0000     4
56.0000     4
43.0000     4
62.0000     4
61.0000     4
1.0000      3
8.0000      3
63.0000     3
41.0000     3
15.0000     2
71.0000     2
20.0000     2
65.0000     2
4.0000      2
6.0000      2
2.0000      1
0.9167      1
14.0000     1
11.0000     1
76.0000     1
59.0000     1
80.0000     1
28.5000     1
32.5000     1
70.0000     1
45.5000     1
13.0000     1
67.0000     1
12.0000     1
0.8333      1
0.6667      1
7.0000      1
3.0000      1
Name: count, dtype: int64

In [ ]:
total_rows = len(df)
total_rows

499

In [ ]:
df.columns

Index(['survived', 'name', 'sex', 'age'], dtype='object')

In [ ]:
# percentage version
total_rows = len(df)

# Menghitung dan menampilkan persentase missing values di setiap kolom satu per satu
for column in df.columns:
    missing_count = df[column].isna().sum()
    missing_percentage = (missing_count / total_rows) * 100
    print(f"Column '{column}' Has {missing_count} missing values ({missing_percentage:.2f}%)") # .2f means 2 decimal

Column 'survived' Has 0 missing values (0.00%)
Column 'name' Has 0 missing values (0.00%)
Column 'sex' Has 0 missing values (0.00%)
Column 'age' Has 49 missing values (9.82%)


In [ ]:
# Handling missing value for EDA, without splitting
for column in df.columns:
    if df[column].dtype == 'object':
        # Jika kolom bertipe object, isi dengan mode
        df[column].fillna(df[column].mode()[0], inplace=True)
    else: # karena tipe data hanya object dan numerik saja tidak ada yg lain, maka pakai else berikut
        # Jika kolom bertipe numerik, isi dengan median
        df[column].fillna(df[column].median(), inplace=True)

<ipython-input-45-f2ed553c45d2>:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[column].fillna(df[column].median(), inplace=True)
<ipython-input-45-f2ed553c45d2>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[column].fillna(df[column].median(), inplace=True)
<ipython-input-45-f2ed553c45d2>:5: FutureWarning: A value is trying to be set

In [ ]:
df.isna().sum()

survived    0
name        0
sex         0
age         0
dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 499 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  499 non-null    int64  
 1   name      499 non-null    object 
 2   sex       499 non-null    object 
 3   age       499 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 19.5+ KB
